aws/src/main/java/org/apache/iceberg/aws/s3/signer/S3V4RestSignerClient.java

```diff
@@ -0,0 +1,328 @@
+      SdkHttpFullRequest.Builder mutableRequest, Aws4PresignerParams signerParams) {
+    return UNSIGNED_PAYLOAD;
+  }
+
+  @Override
+  public SdkHttpFullRequest presign(
+      SdkHttpFullRequest request, ExecutionAttributes executionAttributes) {
+    throw new UnsupportedOperationException("Pre-signing not allowed.");
+  }
+
+  @Override
+  public SdkHttpFullRequest sign(
+      SdkHttpFullRequest request, ExecutionAttributes executionAttributes) {
+    AwsS3V4SignerParams signerParams =
+        extractSignerParams(AwsS3V4SignerParams.builder(), executionAttributes).build();
+
+    S3SignRequest remoteSigningRequest =
+        ImmutableS3SignRequest.builder()
+            .method(request.method().name())
+            .region(signerParams.signingRegion().id())
+            .uri(request.getUri())
+            .headers(request.headers())
+            .properties(requestPropertiesSupplier().get())
+            .build();
+
+    Key cacheKey = Key.from(remoteSigningRequest);
+    SignedComponent cachedSignedComponent = SIGNED_COMPONENT_CACHE.getIfPresent(cacheKey);
+
+    if (cachedSignedComponent != null) {
+      // The SdkHttpFullRequest Builder appends the raw path from the input URI in .uri(),
+      // so we need to clear the current path from the request
+      SdkHttpFullRequest.Builder mutableRequest = request.toBuilder();
+      mutableRequest.encodedPath("");
+      mutableRequest.uri(cachedSignedComponent.signedURI());
+      reconstructHeaders(cachedSignedComponent.headers(), mutableRequest);
+
+      return mutableRequest.build();
+    }
+
+    S3SignResponse s3SignResponse =
+        httpClient()
+            .post(
+                endpoint(),
+                remoteSigningRequest,
+                S3SignResponse.class,
+                () -> OAuth2Util.authHeaders(authSession().token()),
+                ErrorHandlers.defaultErrorHandler());
+
+    if (canBeCached(s3SignResponse)) {
+      SIGNED_COMPONENT_CACHE.put(
+          cacheKey,
+          ImmutableSignedComponent.builder()
+              .headers(s3SignResponse.headers())
+              .signedURI(s3SignResponse.uri())
+              .build());
+    }
+
+    // Update URI and append signed and unsigned headers
+    // The SdkHttpFullRequest Builder appends the raw path from the input URI in .uri(),
+    // so we need to clear the current path from the request
+    SdkHttpFullRequest.Builder mutableRequest = request.toBuilder();
+    mutableRequest.encodedPath("");
+    mutableRequest.uri(s3SignResponse.uri());
+    reconstructHeaders(s3SignResponse.headers(), mutableRequest);
+
+    return mutableRequest.build();
+  }
+
+  private void reconstructHeaders(
+      Map<String, List<String>> signedAndUnsignedHeaders,
+      SdkHttpFullRequest.Builder mutableRequest) {
+    Map<String, List<String>> headers = Maps.newHashMap(signedAndUnsignedHeaders);
+    // we need to remove the Cache-Control header that is being sent by the server
```

Comment:
Why does this override the cache control from the service?

Reply:
the cache-control header is only meant for the S3 Signer client, so we shouldn't propagate it further. See also my comment [here](https://github.com/apache/iceberg/pull/6169#discussion_r1031590899) about why the cache-control header is in the headers of the response